pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+			};
+
+			let valid_tx = |provide| {
+				ValidTransaction::with_tag_prefix("settle-otc-with-router")
+					.priority(UNSIGNED_TXS_PRIORITY)
+					.and_provides([&provide])
+					.longevity(3)
+					.propagate(false)
+					.build()
+			};
+
+			match call {
+				Call::settle_otc_order { .. } => valid_tx(b"settle_otc_order".to_vec()),
+				_ => InvalidTransaction::Call.into(),
+			}
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A trade has been executed
+		Executed {
+			otc_id: OrderId,
+			otc_asset_in: AssetIdOf<T>,
+			otc_asset_out: AssetIdOf<T>,
+			otc_amount_in: Balance,
+			otc_amount_out: Balance,
+			trade_amount_in: Balance,
+			trade_amount_out: Balance,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Otc order not found
+		OrderNotFound,
+		/// Execution requirements not met
+		InvalidConditions,
+		/// Trade amount higher than necessary
+		TradeAmountTooHigh,
+		/// Trade amount lower than necessary
+		TradeAmountTooLow,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Close an existing OTC arbitrage opportunity.
+		///
+		/// Executes a trade between an OTC order and some route.
+		/// If the OTC order is partially fillable, the extrinsic fails if the existing arbitrage
+		/// opportunity is not closed after the trade.
+		/// If the OTC order is not partially fillable, fails if there is no profit after the trade.
+		///
+		/// `Origin` calling this extrinsic is not paying or receiving anything.
+		///
+		/// The profit made by closing the arbitrage is transferred to `FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: Unsigned origin.
+		/// - `otc_id`: ID of the OTC order with existing arbitrage opportunity.
+		/// - `amount`: Amount necessary to clone the arb.
+		/// - `route`: The route we trade against. Required for the fee calculation.
+		///
+		/// Emits `Executed` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::settle_otc_order()
+			.saturating_add(<T as Config>::RouterWeightInfo::sell_weight(route))
+		.saturating_add(<T as Config>::RouterWeightInfo::get_route_weight())
+		.saturating_add(<T as Config>::RouterWeightInfo::calculate_spot_price_weight(route))
+		.saturating_add(<T as pallet_otc::Config>::WeightInfo::fill_order().max(<T as pallet_otc::Config>::WeightInfo::partial_fill_order()))
+		)]
+		pub fn settle_otc_order(
+			origin: OriginFor<T>,
+			otc_id: OrderId,
+			amount: Balance,
+			route: Vec<Trade<AssetIdOf<T>>>,
+		) -> DispatchResult {
+			ensure_none(origin)?;
+
+			Self::settle_otc(otc_id, amount, route)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn account_id() -> T::AccountId {
+		PALLET_ID.into_account_truncating()
+	}
+
+	/// Ensure that the profit is more than some minimum amount.
+	fn ensure_min_profit(asset: T::AssetId, _profit: Balance) -> DispatchResult {
+		let _min_amount = <T as Config>::ExistentialDeposits::get(&asset)
```

Comment:
this approach doesn't really work for insufficient assets as their existential deposit is 1

Reply:
We get EDs from ED provider, this was supposed to handle cases when ED == 1 (insufficient assets)